File: backend/mcp-code-intelligence-python/src/mcp_code_intel/memory/graph_analytics.py
"""GraphAnalytics — centrality, hubs, connected components analysis."""

from __future__ import annotations

from collections import deque

AdjMap = dict[int, set[int]]


def degree_centrality(adj: AdjMap) -> dict[int, float]:
    """Degree centrality — normalized count of connections per node."""
    if not adj:
        return {}
    max_degree = max((len(n) for n in adj.values()), default=0)
    if max_degree == 0:
        return {k: 0.0 for k in adj}
    return {node: len(neighbors) / max_degree for node, neighbors in adj.items()}


def find_hubs(adj: AdjMap, min_degree: int = 3) -> list[int]:
    """Find hub nodes — nodes with degree above threshold."""
    return [node for node, neighbors in adj.items() if len(neighbors) >= min_degree]


def find_isolated(adj: AdjMap) -> list[int]:
    """Find isolated nodes — nodes with no connections."""
    return [node for node, neighbors in adj.items() if len(neighbors) == 0]


def connected_components(adj: AdjMap) -> list[set[int]]:
    """Connected components using BFS."""
    visited: set[int] = set()
    components: list[set[int]] = []
    for node in adj:
        if node in visited:
            continue
        component: set[int] = set()
        queue: deque[int] = deque([node])
        while queue:
            current = queue.popleft()
            if current in visited:
                continue
            visited.add(current)
            component.add(current)
            for neighbor in adj.get(current, set()):
                if neighbor not in visited:
                    queue.append(neighbor)
        components.append(component)
    return components


def density(node_count: int, edge_count: int) -> float:
    """Graph density — ratio of actual edges to possible edges."""
    if node_count <= 1:
        return 0.0
    max_edges = node_count * (node_count - 1)
    return edge_count / max_edges

File: backend/mcp-code-intelligence-nodejs/src/memory/graph-analytics.ts
/**
 * GraphAnalytics — centrality, hubs, connected components analysis.
 */

type AdjMap = Map<number, Set<number>>;

/** Degree centrality — normalized count of connections per node. */
export function degreeCentrality(adj: AdjMap): Map<number, number> {
  if (adj.size === 0) return new Map();
  let maxDegree = 0;
  for (const neighbors of adj.values()) {
    if (neighbors.size > maxDegree) maxDegree = neighbors.size;
  }
  if (maxDegree === 0) return new Map([...adj.keys()].map(k => [k, 0]));
  const result = new Map<number, number>();
  for (const [node, neighbors] of adj) {
    result.set(node, neighbors.size / maxDegree);
  }
  return result;
}

/** Find hub nodes — nodes with degree above threshold. */
export function findHubs(adj: AdjMap, minDegree = 3): number[] {
  const hubs: number[] = [];
  for (const [node, neighbors] of adj) {
    if (neighbors.size >= minDegree) hubs.push(node);
  }
  return hubs;
}

/** Find isolated nodes — nodes with no connections. */
export function findIsolated(adj: AdjMap): number[] {
  const isolated: number[] = [];
  for (const [node, neighbors] of adj) {
    if (neighbors.size === 0) isolated.push(node);
  }
  return isolated;
}

/** Connected components using BFS. */
export function connectedComponents(adj: AdjMap): Set<number>[] {
  const visited = new Set<number>();
  const components: Set<number>[] = [];
  for (const node of adj.keys()) {
    if (visited.has(node)) continue;
    const component = new Set<number>();
    const queue: number[] = [node];
    while (queue.length > 0) {
      const current = queue.shift()!;
      if (visited.has(current)) continue;
      visited.add(current);
      component.add(current);
      for (const neighbor of adj.get(current) ?? []) {
        if (!visited.has(neighbor)) queue.push(neighbor);
      }
    }
    components.push(component);
  }
  return components;
}

/** Graph density — ratio of actual edges to possible edges. */
export function density(nodeCount: number, edgeCount: number): number {
  if (nodeCount <= 1) return 0;
  const maxEdges = nodeCount * (nodeCount - 1);
  return edgeCount / maxEdges;
}
